### PR TITLE
[SPARK-42525][SQL] Collapse two adjacent windows with semantically-same partition/order

### PR DESCRIPTION
**Pull request by zml1206**
### What changes were proposed in this pull request?
Extend the CollapseWindow rule to collapse Window nodes, when one window in subquery.

### Why are the changes needed?

```
select a, b, c, row_number() over (partition by a order by b) as d from
( select a, b, rank() over (partition by a order by b) as c from t1) t2

== Optimized Logical Plan ==
before
Window [row_number() windowspecdefinition(a#11, b#12 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS d#26], [a#11], [b#12 ASC NULLS FIRST]
+- Window [rank(b#12) windowspecdefinition(a#11, b#12 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS c#25], [a#11], [b#12 ASC NULLS FIRST]
   +- InMemoryRelation [a#11, b#12], StorageLevel(disk, memory, deserialized, 1 replicas)
         +- *(1) Project [_1#6 AS a#11, _2#7 AS b#12]
            +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, scala.Tuple2, true]))._1 AS _1#6, knownnotnull(assertnotnull(input[0, scala.Tuple2, true]))._2 AS _2#7]
               +- *(1) MapElements org.apache.spark.sql.DataFrameSuite$$Lambda$1517/1628848368@3a479fda, obj#5: scala.Tuple2
                  +- *(1) DeserializeToObject staticinvoke(class java.lang.Long, ObjectType(class java.lang.Long), valueOf, id#0L, true, false, true), obj#4: java.lang.Long
                     +- *(1) Range (0, 10, step=1, splits=2)

after
Window [rank(b#12) windowspecdefinition(a#11, b#12 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS c#25, row_number() windowspecdefinition(a#11, b#12 ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS d#26], [a#11], [b#12 ASC NULLS FIRST]
+- InMemoryRelation [a#11, b#12], StorageLevel(disk, memory, deserialized, 1 replicas)
      +- *(1) Project [_1#6 AS a#11, _2#7 AS b#12]
         +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, scala.Tuple2, true]))._1 AS _1#6, knownnotnull(assertnotnull(input[0, scala.Tuple2, true]))._2 AS _2#7]
            +- *(1) MapElements org.apache.spark.sql.DataFrameSuite$$Lambda$1518/1928028672@4d7a64ca, obj#5: scala.Tuple2
               +- *(1) DeserializeToObject staticinvoke(class java.lang.Long, ObjectType(class java.lang.Long), valueOf, id#0L, true, false, true), obj#4: java.lang.Long
                  +- *(1) Range (0, 10, step=1, splits=2)
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

**Comment by zml1206** [comment]
cc @cloud-fan @wangyum

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -3592,6 +3592,34 @@ class DataFrameSuite extends QueryTest
     val df = Seq("0.5944910").toDF("a")
     checkAnswer(df.selectExpr("cast(a as decimal(7,7)) div 100"), Row(0))
   }
+
+  test("SPARK-42525: collapse two adjacent windows with the same partition/order in subquery") {
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:3596]: This is because their qualifiers are different. Please add a test to `CollapseWindowSuite`:
```scala
val query = testRelation
  .window(Seq(min(a).as("_we0")), Seq(c.withQualifier(Seq("0"))), Seq(c.asc))
  .select($"a", $"b", $"c", $"_we0" as "min_a")
  .window(Seq(max(a).as("_we1")), Seq(c.withQualifier(Seq("1"))), Seq(c.asc))
  .select($"a", $"b", $"c", $"min_a", $"_we1" as "max_a")
  .analyze

val optimized = Optimize.execute(query)

val correctAnswer = testRelation
  .window(Seq(min(a).as("_we0"), max(a).as("_we1")), Seq(c), Seq(c.asc))
  .select(a, b, c, $"_we0" as "min_a", $"_we1" as "max_a")
  .analyze

comparePlans(optimized, correctAnswer)
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 2]
```diff
@@ -3592,6 +3592,34 @@ class DataFrameSuite extends QueryTest
...
+        |from t1
+        |""".stripMargin
+    )
+    val captured = new ByteArrayOutputStream()
+    Console.withOut(captured) {
+      df2.explain()
+    }
+    checkAnswer(df2, df3)
+    val output = captured.toString
+    assert(output.split("\\+- Window").size == 2)
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:3621]: Move this test to `DataFrameWindowFunctionsSuite`
```scala
withTempView("t1") {
  Seq((1, 1), (2, 2)).toDF("a", "b").createOrReplaceTempView("t1")
  val df = sql(
    """
      |SELECT a, b, rk, row_number() OVER (PARTITION BY a ORDER BY b) AS rn
      |FROM   (SELECT a, b, rank() OVER (PARTITION BY a ORDER BY b) AS rk
      |        FROM t1) t2
      |""".stripMargin)

  val windows = df.queryExecution.optimizedPlan.collect { case w: LogicalWindow => w }
  assert(windows.size === 1)
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala [hunk 3]
```diff
@@ -532,10 +532,15 @@ class DataFrameWindowFunctionsSuite extends QueryTest
   test("window function with aggregator") {
     val agg = udaf(new Aggregator[(Long, Long), Long, Long] {
       def zero: Long = 0L
+
```
Comment (wangyum) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala:535]: Please revert these changes.

**Comment by wangyum** [comment]
@zml1206 Could you update the PR title to `[SPARK-42525][SQL] Collapse ...`?

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master.

**Comment by cloud-fan** [comment]
the change LGTM but the PR title is a bit confusing. How is it related to subquery?

**Comment by zml1206** [comment]
> the change LGTM but the PR title is a bit confusing. How is it related to subquery?

subquery is one of the cases where the qualifiers are different，it's really confusing, is there anything else I can do to modify it

**Comment by cloud-fan** [comment]
`... with semantically-same partition/order`
